core/src/main/java/io/grpc/internal/AbstractClientStream.java

```diff
@@ -116,6 +117,10 @@ protected void inboundHeadersReceived(Metadata.Headers headers) {
       log.log(Level.INFO, "Received headers on closed stream {0} {1}",
           new Object[]{id(), headers});
     }
+    if (headers.containsKey(ChannelImpl.MESSAGE_ENCODING_KEY)) {
+      String messageEncoding = headers.get(ChannelImpl.MESSAGE_ENCODING_KEY);
+      setDecompressor(messageEncoding);
```

Comment:
Since @nmittler  LGTM'd this, I am going to go ahead an submit.  If there are lingering issues comment on this PR and I can address them in a followup PR.